File: hands-on/ho10/graph_solver.py
import threading
import time

def compute(op1, op2, operation, result_list, index):
    """Funzione per le operazioni elementari"""
    if operation == '*':
        res = op1 * op2
    elif operation == '+':
        res = op1 + op2
    elif operation == '-':
        res = op1 - op2
    else:
        res = 0
    
    print(f"[Thread {chr(65+index)}] Eseguo: {op1} {operation} {op2} = {res}")
    result_list[index] = res

def main():
    print("Risoluzione equazione: y = (2 * 6) + (1 + 4) * (5 - 2)")
    print("Avvio dei thread per le operazioni indipendenti...\n")

    # Lista per i risultati (per catturare i valori di ritorno)
    results = [None] * 3
    
    # Lancio i primi 3 thread in parallelo
    tA = threading.Thread(target=compute, args=(2, 6, '*', results, 0))
    tB = threading.Thread(target=compute, args=(1, 4, '+', results, 1))
    tC = threading.Thread(target=compute, args=(5, 2, '-', results, 2))

    tA.start()
    tB.start()
    tC.start()

    # Sincronizzazione: attendo i risultati (Join)
    tA.join()
    tB.join()
    tC.join()

    print("\nRisultati parziali ottenuti. Calcolo i nodi dipendenti...")

    # Nodo D = B * C
    resD = results[1] * results[2]
    print(f"[Main] Nodo D (B * C): {results[1]} * {results[2]} = {resD}")

    # Nodo E = A + D (Risultato finale)
    final_y = results[0] + resD
    print(f"[Main] Risultato Finale y (A + D): {results[0]} + {resD} = {final_y}")

if __name__ == "__main__":
    main()

File: hands-on/ho10/graph_solver.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

/* Struttura per passare gli operandi ai thread */
typedef struct {
    int op1;
    int op2;
    char operation;
} thread_args_t;

/* Funzioni per le operazioni elementari */
void* compute(void* arg) {
    thread_args_t* args = (thread_args_t*)arg;
    int* result = malloc(sizeof(int));

    switch (args->operation) {
        case '*': *result = args->op1 * args->op2; break;
        case '+': *result = args->op1 + args->op2; break;
        case '-': *result = args->op1 - args->op2; break;
        default: *result = 0; break;
    }

    printf("[Thread %c] Eseguo: %d %c %d = %d\n", 
           (args->operation == '*' ? 'A' : (args->operation == '+' ? 'B' : 'C')),
           args->op1, args->operation, args->op2, *result);

    pthread_exit(result);
}

int main() {
    pthread_t tA, tB, tC;
    thread_args_t argsA = {2, 6, '*'};  /* Nodo A: 2 * 6 */
    thread_args_t argsB = {1, 4, '+'};  /* Nodo B: 1 + 4 */
    thread_args_t argsC = {5, 2, '-'};  /* Nodo C: 5 - 2 */

    int *resA, *resB, *resC;

    printf("Risoluzione equazione: y = (2 * 6) + (1 + 4) * (5 - 2)\n");
    printf("Avvio dei thread per le operazioni indipendenti...\n\n");

    /* Lancio i primi 3 thread in parallelo */
    pthread_create(&tA, NULL, compute, &argsA);
    pthread_create(&tB, NULL, compute, &argsB);
    pthread_create(&tC, NULL, compute, &argsC);

    /* Sincronizzazione: attendo i risultati */
    pthread_join(tA, (void**)&resA);
    pthread_join(tB, (void**)&resB);
    pthread_join(tC, (void**)&resC);

    printf("\nRisultati parziali ottenuti. Calcolo i nodi dipendenti...\n");

    /* Nodo D = B * C (Dipende dai thread B e C) */
    int resD = (*resB) * (*resC);
    printf("[Main] Nodo D (B * C): %d * %d = %d\n", *resB, *resC, resD);

    /* Nodo E = A + D (Risultato finale) */
    int final_y = (*resA) + resD;
    printf("[Main] Risultato Finale y (A + D): %d + %d = %2d\n", *resA, resD, final_y);

    /* Pulizia memoria */
    free(resA);
    free(resB);
    free(resC);

    return 0;
}
